src/herder/TxSetFrame.cpp

```diff
@@ -31,55 +31,135 @@
-    return mHash;
-}
+    if (txSet.v() != 1)
+    {
+        CLOG_DEBUG(Herder, "Got bad txSet: unsupported version {}", txSet.v());
+        return false;
+    }
+    auto const& txSetV1 = txSet.v1TxSet();
+    if (txSetV1.phases.size() != 1)
+    {
+        CLOG_DEBUG(Herder, "Got bad txSet: exactly 1 phase is expected, got {}",
+                   txSetV1.phases.size());
+        return false;
+    }
 
-Hash const&
-TxSetFrame::previousLedgerHash() const
-{
-    return mPreviousLedgerHash;
+    auto const& phase = txSetV1.phases[0];
+    if (phase.v() != 0)
+    {
+        CLOG_DEBUG(Herder, "Got bad txSet: unsupported phase version {}",
+                   phase.v());
+        return false;
+    }
+    auto bidIsFeeComponents =
+        std::count_if(phase.v0Components().begin(), phase.v0Components().end(),
+                      [](auto const& component) {
+                          return !component.txsMaybeDiscountedFee().baseFee;
+                      });
+    if (bidIsFeeComponents > 1)
+    {
+        CLOG_DEBUG(Herder, "Got bad txSet: more than 1 BID_IS_FEE component {}",
+                   bidIsFeeComponents);
+        return false;
+    }
+
+    bool componentsNormalized =
+        std::is_sorted(phase.v0Components().begin(), phase.v0Components().end(),
+                       [](auto const& c1, auto const& c2) {
+                           if (!c1.txsMaybeDiscountedFee().baseFee ||
+                               !c2.txsMaybeDiscountedFee().baseFee)
+                           {
+                               return !c1.txsMaybeDiscountedFee().baseFee &&
+                                      c2.txsMaybeDiscountedFee().baseFee;
+                           }
+                           return *c1.txsMaybeDiscountedFee().baseFee <
+                                  *c2.txsMaybeDiscountedFee().baseFee;
+                       });
+    if (!componentsNormalized)
+    {
+        CLOG_DEBUG(Herder, "Got bad txSet: incorrect component order");
+        return false;
+    }
+
+    bool componentBaseFeesUnique =
+        std::adjacent_find(phase.v0Components().begin(),
+                           phase.v0Components().end(),
+                           [](auto const& c1, auto const& c2) {
+                               if (!c1.txsMaybeDiscountedFee().baseFee ||
+                                   !c2.txsMaybeDiscountedFee().baseFee)
+                               {
+                                   return !c1.txsMaybeDiscountedFee().baseFee &&
+                                          !c2.txsMaybeDiscountedFee().baseFee;
+                               }
+                               return *c1.txsMaybeDiscountedFee().baseFee ==
+                                      *c2.txsMaybeDiscountedFee().baseFee;
+                           }) == phase.v0Components().end();
+    if (!componentBaseFeesUnique)
+    {
+        CLOG_DEBUG(Herder, "Got bad txSet: duplicate component base fees");
+        return false;
+    }
+
+    return true;
```

Comment:
there is a missing check that `txs` cannot be empty (this is also missing from CAP-0042), otherwise we have an infinite number of ways to represent an empty set.

The only way should be to have `v0Components` empty.